### ISSUE TITLE
Test failure on some non-x86 systems

### ISSUE THREAD

**Issue by QuLogic** [eddelbuettel/nanotime#34]
When running tests on non-x86 systems, some of them will fail. [This build](https://koji.fedoraproject.org/koji/taskinfo?taskID=27499419) runs on a variety of systems and fails with tests on some of them. If you check the ppc64, s390x, or armv7hl builds, they all fail a single test. Unfortunately, I don't have interactive access to s390x or armv7hl systems, but I can check the result on ppc64.

There seems to be some problem in `test_format_default`:
```
test_format_default: FAILURE !! (check number 2)
Error in checkEquals(format(nanotime("1680-07-17T00:00:01.000000000+00:00")),  : 
  1 string mismatch
```
If I try writing it out manually, the result is:
```r
> format(nanotime("2120-01-01T00:00:59.987654321+00:00"))
[1] "2120-01-01T00:00:59.987654833+00:00"
```
so there seems to be some kind of floating point issue in parsing things.

**Comment by eddelbuettel** [comment]
That just looks like a lack of precision beyond the '4', ie `.987654` is agreed upon. 

I am guessing here but those system may just not have the nanosecond resolution the C++ `chrono` type _can_ provide _if_ the compiler / host / os support it.

In short, that looks like a CCTZ issue on that platform.  You could report it as a short (five or so lines should do) C++ program at the [CCTZ GitHub repo](https://github.com/google/cctz).

For `nanotime`, I suggest to relax the comparison precision on those platforms.

**Comment by QuLogic** [comment]
The problem appears to be [here](https://github.com/eddelbuettel/nanotime/blob/master/R/nanotime.R#L127):
```R
> d <- RcppCCTZ::parseDouble("2120-01-01T00:00:59.987654321+00:00", fmt="%Y-%m-%dT%H:%M:%E9S%Ez", tz="UTC")
> d
           [,1]      [,2]
[1,] 4733510459 987654321

> as.integer64(d[,1]) * 1e9
integer64
[1] 4733510459000000512

> as.integer64(d[,1]) * as.integer64(1e9)
integer64
[1] 4733510459000000000
```
I don't know if that qualifies as a bug in bit64 or just the usual floating-point issues, but it appears easily fixable here.

**Comment by QuLogic** [comment]
Also working:
```R
> as.integer64(d[,1]) * 1000000000L
integer64
[1] 4733510459000000000

> as.integer64(d[,1]) * 1e9L
integer64
[1] 4733510459000000000
```

**Comment by eddelbuettel** [comment]
Nice catch, clearly on my side.  That point is where we can loose precision between 53-bit doubles and the 64 bit ints we don't "really" have in R.  But I like that idea a bit.  @lsilvest do you see anything wrong this, or should be try this?

**Comment by eddelbuettel** [comment]
Looking your example again:

```r
R> d <- RcppCCTZ::parseDouble("2120-01-01T00:00:59.987654321+00:00", 
+                             fmt="%Y-%m-%dT%H:%M:%E9S%Ez", tz="UTC")        
R> d 
           [,1]      [,2]           
[1,] 4733510459 987654321       
R>  
```

So far, so good.  We specified something at 18 digits precision, and we got 18 digits back.

But in the rest, you are demonstrating that doubles only have 15 digits, which is what we know.
Also my amd64 seems to be kinder:

```r
R> suppressMessages(library(bit64))
R> as.integer64(d[,1]) * 1000000000L
integer64
[1] 4733510459000000000
R> as.integer64(d[,1]) * 1e9L
integer64
[1] 4733510459000000000
R> as.integer64(d[,1]) * 1.0e9
integer64
[1] 4733510459000000000
R> as.integer64(d[,1]) * 1e9
integer64
[1] 4733510459000000000
R> 
```

Here I can scale out clearly because `bit64` helps.  But in general we will loose precision when 
going to R's representation so I still think we should just not do those tests on platforms which are
less generous to us than amd64.

**Comment by QuLogic** [comment]
IIRC, x86 has some extended registers (80-bit) that can be a bit more forgiving in some circumstances. I'm not totally familiar with the ppc64 system, so I do find it odd that LE is fine but BE is not.

In any case, when working with large numbers like this, it seems best to avoid fp math altogether.

**Comment by eddelbuettel** [comment]
Yes.  `nanotime` is a compromise as R _does not have native `int64_t`_ so corner cases appear.

But it is very useful on some platforms, and we made the package way more robust (via a S4 rewrite, thanks to @lsilvest ) than it was before, and it fills a need.   

Let me know if you need upstream help to turn some tests off on those platforms.  On Debian, we tend to build everything everywhere (ie every package on every platforms where it compiles) and we can aid that here by suppressing some tests that would puke on some of those systems.

**Comment by QuLogic** [comment]
I'm not quite sure I understand why it's preferable to change tolerances or disable tests rather than add the integer `L` suffix (or `as.integer64`) on the few used constants? 1e9 is within a 32-bit integer so the lack of a 64-bit integer type is not super important as far as the constant is concerned (and it's multiplied by a `bit64::integer64` anyway.) And using `as.integer(1000000000)` is already done for the ['reverse' operation in `format.nanotime`](https://github.com/eddelbuettel/nanotime/blob/master/R/nanotime.R#L209).

**Comment by eddelbuettel** [comment]
Slow down.  No tests are disabled. 

I showed your issue was not reproducable at my end, hence "ball in your court".  And I am not sure why you are arguing endlessly with me. Propose a clean, small, ... pull request, ideally with a new test and folks may listen.

Anyway, try [this branch with two new casts](https://github.com/eddelbuettel/nanotime/tree/feature/int64_cast).

**Comment by eddelbuettel** [comment]
Thanks for reporting back.  It has since been released to CRAN as new version 0.2.1, see https://cloud.r-project.org/web/packages/nanotime/index.html
